core/src/main/java/com/linecorp/armeria/internal/PathAndQuery.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal;
+
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.Flags;
+
+/**
+ * A parser of the raw path and query components of an HTTP path. Performs validation and allows caching of
+ * results.
+ */
+public final class PathAndQuery {
+
+    /**
+     * According to RFC 3986 section 3.3, path can contain a colon, except the first segment.
+     *
+     * <p>Should allow the asterisk character in the path, query, or fragment components of a URL(RFC2396).
+     * @see <a href="https://tools.ietf.org/html/rfc3986#section-3.3">RFC 3986, section 3.3</a>
+     */
+    private static final Pattern PROHIBITED_PATH_PATTERN =
+            Pattern.compile("^/[^/]*:[^/]*/|[|<>\\\\]|/\\.\\.|\\.\\.$|\\.\\./");
+
+    private static final Pattern CONSECUTIVE_SLASHES_PATTERN = Pattern.compile("/{2,}");
+
+    private static final Cache<String, PathAndQuery> CACHE =
```

Comment:
Not sure useful or not but how about exposing this cache stats to e.g. prometheus?